src/v/datalake/schema_registry.cc

```diff
@@ -0,0 +1,46 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
+ *
+ * Licensed as a Redpanda Enterprise file under the Redpanda Community
+ * License (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/redpanda-data/redpanda/blob/master/licenses/rcl.md
+ */
+#include "datalake/schema_registry.h"
+
+#include "model/record.h"
+
+namespace datalake {
+std::optional<uint32_t> get_value_schema_id(const model::record& record) {
+    const iobuf& value = record.value();
+    if (value.size_bytes() < 5) {
+        return std::nullopt;
+    }
+    iobuf_const_parser parser(value);
+    auto magic = parser.consume_type<uint8_t>();
+    if (magic != 0) {
+        return std::nullopt;
+    }
+    auto id = parser.consume_type<uint32_t>();
+    return id;
```

Comment:
This might not be sufficient for protobuf, which uses [zig-zag encoded offsets] to get correct message if the Subject Name Strategy isn't TopicName. (https://github.com/jcipar/redpanda/blob/jcipar/datalake-schema-registry/src/v/pandaproxy/schema_registry/validation.cc#L77).

It may be worth extracting and reusing some of the code in [the existing validation](https://github.com/jcipar/redpanda/blob/jcipar/datalake-schema-registry/src/v/pandaproxy/schema_registry/validation.cc#L199-L220)